### DISCUSSION TITLE
Cluster setup with Quarkus in AWS ECS - missing jdbcping - what solution for infinispan

### DISCUSSION THREAD

**Discussion by squalou** [keycloak/keycloak#11859] [Archive]
Hi,
I'm moving to keycloak 17 quarkus.
My production environments are AWS ECS  (not kubernetes, and won't be anytime soon)

Issue is ... JDBC_PING seems to have been removed, and I can't find any other mean to make cluster members discover themselves.

Let me explain.

If I set KC_CACHE_STACK=ec2 and all the required environment variables to point to a S3 bucket,
when a container starts it creates a file there BUT records in it its "local" docker network IP. (172.x.x.x.)

when another container starts, it most probably will start on *another* ec2 instance of the ECS cluster, so ... the internal docker IP is completely useless. 
(moreover I use static port mapping for other reasons too long to be explained here, as a result 2 containers cannot start on the same ec2 instance anyway))

With Keycloak 15 I used to have a dedicated startup script that managed to get the "outer" host IP, and then use it to feed JDBC_PING mechanism.

Now ... with S3 I have no idea what to do. 

(I could maybe customize kc.sh, edit the s3 file after a node starts, change the IP field, but ... I don't even know how to find the *proper* s3 file given its weird name ... this all looks like ugly workarounds I have no idea how to make work properly, let alone maintain)

Additionally, s3 file contains IP:port with port = 7800... I'm not even sure what I should map this port to. (I mean, is i always 7800 that is looked for)

Any idea welcome.

JDBC_PING dearly missing and custom startup scripts too 

for the records, I had a look there too : https://github.com/keycloak/keycloak/discussions/10177

**Comment by Arulaln-AR** [comment]
Hi,

Have you checked the article "https://gist.github.com/xgp/768eea11f92806b9c83f95902f7f8f80".

It has some steps and it is working locally when i tried to set keycloak cluster.

**Comment by whbogado** [comment]
Hi @squalou,

To S3_PING to work correctly you have to setup your task to **awsvpc** or **host** network mode. I have only tested awsvpc mode but it should also work in host mode. Although bridge is more flexible, unfortunately I couldn't find a reasonable workaround to make it work. Because S3_PING manages the creation and deletion of the bucket files you can't tamper with them reliably.

In awsvpc mode each host will have 1 or more network interface (ENI) attached an mapped to the container, this way they can find each other, even if they run in different EC2 instances and different AZs. Looking at the source code for S3_PING and the logs I can see this is related to the GMS discovery protocol, not directly to S3_PING implementation.

With this setup the container IP will be from the VPC network. You can't have dynamic ports but you can map host port 7800 to container port 7800. As long as security groups and instance roles are correct it should work.
